text/0000-fn-delegation.md

```diff
@@ -0,0 +1,712 @@
+|-------------------:|---------------:|
+|                  1 |           5124 |
+|                  2 |            279 |
+|                  3 |             97 |
+|                  4 |             72 |
+|                  5 |             49 |
+|                 10 |             29 |
+|                  6 |             28 |
+|                  8 |             17 |
+|                  7 |             16 |
+|                 14 |              8 |
+|                 12 |              4 |
+|                  9 |              3 |
+|                 13 |              2 |
+|                ... |            ... |
+
+</details>
+
+The feature should not be hard to support compared to the gains, unless there are some issues with
+cloning `target_expr_template`'s IR or something. In any case, the mechanism will be necessary if
+we are going to support higher level sugar like delegation all functions in a trait impl at once
+with "glob delegation".
+
+## Parent context
+
+Functions delegation should be supported inside both trait impls and inherent impls.
+Only about 60% of delegation instances correspond to items in trait impls.
+
+It should also be supported in other contexts too (free functions, default methods in traits),
+because why not - all these cases fit into our code generation scheme (see below) and occur in
+practice.
+Free function support can be sacrificed if it stops fitting into the scheme, though.
+A typical free function delegation case collected from stdarch looks like this.
+```rust
+#[target_feature(enable = "some_feature")]
+pub fn foo(arg: u32) { submodule::foo(arg) }
+
+// In delegation form
+#[target_feature(enable = "some_feature")]
+pub delegation_keyword submodule::foo;
+```
+Without the attribute difference such "delegation" could just be replaced with a simple reexport
+`pub use submodule::foo;`.
+
+| caller parent             |   count |
+|:--------------------------|--------:|
+| CallerParent.TraitImpl    |    3124 |
+| CallerParent.InherentImpl |    1601 |
+| CallerParent.Other        |     462 |
+| CallerParent.Trait        |     107 |
+
+## First argument transformation aka "target expression"
+
+First argument (often `self`) should supports arbitrary transformations, rather than just
+projection to a field (`self.field`).
+In other words, the target expression should be an arbitrary expression.
+
+| arg0 preproc         |   count |                     |
+|:---------------------|---------|--------------------:|
+| Arg0Preproc.No       |    1734 | `arg0`              |
+| Arg0Preproc.Field    |    1582 | `arg0.field`        |
+| Arg0Preproc.Other    |    1270 | anything else       |
+| Arg0Preproc.Getter   |     583 | `arg0.get_field()`  |
+| Arg0Preproc.RefField |     136 | `&(mut) arg0.field` |
+
+`self.field` is not even the most common transformation, usually the first argument is not
+transformed at all, although delegation for static methods and free functions skews the statistics
```

Comment:
What's a "static method"?

Reply:
Yeah, that's just my C++/Java past leaking, the terminology seems so natural I didn't realize it's not often used in Rust since it doesn't use `static` for this purpose.
Grepping the rust-lang/rust repo for "static method" shows that quite a few people stepped into the same issue though.

Reply:
> In Java static methods are statically dispatched while regular methods are dynamically dispatched; that does not apply for us either.

Is that what `static` is supposed to mean? In Java, what affects whether the dispatch is dynamic would be `final` rather than `static`, and `static` not being dynamically dispatched is just a consequence of not having a receiver.

Reply:
> Is that what static is supposed to mean?

No idea, it's just my mental model. I can't otherwise explain why these would be called `static` methods.

But in Java they have to be static *methods* since Java only has methods, not functions. In Rust, calling them any kind of method makes little sense, you can't even use method call syntax to call them.

Reply:
> No idea, it's just my mental model. I can't otherwise explain why these would be called `static` methods.

Historically speaking, Java got the term `static` from C++, and in C++ dynamic dispatch is not the default, requiring a separate keyword (`virtual`); so that can't have been the original motivation for the term.  I have no idea why they picked `static`, but it probably helped that `static` was already a keyword in C, albeit with an unrelated meaning.

(In truth, both C and C++ are guilty of overloading `static` to mean several different things depending on context.)